git-big-picture

```diff
@@ -434,17 +409,7 @@ def parse_filter_options(opts, SETTINGS):
     """
     cmdline_settings = {}
     for setting in SETTINGS:
-        positive, negative = getattr(opts, setting), getattr(opts, 'no_' + setting)
```

Comment:
This fundamentally changes the behavior. Before, when two options like `-C` and `-c` were used , the CLI would barf  informing the user that two options were mutually exclusive. Now, the last option simply wins. I think I prefer the program to be explicit in failure modes rather than silently discarding a supplied option. Do you think  you could bring back the original (and maybe even add  some tests as it even had a custom exit code)?

Reply:
It was a conscious decision to drop it, not driving by saving effort. If we were to start on a green field, I think our options would be:
- a) mark `--foo` and `--no-foo` as mutually exclusive with argparse, have argparse fail to the user with a clean auto-generated message, have argparse exit with code 1
- b) do *not* mark `--foo` and `--no-foo` as mutually exclusive with argparse but handle exclusivity manually, return the same magic number as used previously.
- c) implement "last request wins" approach, have things override each other and be okay with that, actively support it.

(b) is the one that I prefer least, because argparse can do that work for us and I see little value in rescuing that magic exit code number since argparse based applications return exit code 1 for all command line parsing errors. I think that approach is accepted and okay: We're free to use 2 and above for anything else (and we already do). So no gain in manually coding exclusivity outside of argparse in my view.

With we go with (a) or (c) is a design decision, I think. Personally I think we should reserve exclusivity for things that do not control the same flag. I would go with (c) personally but I'm find with (a) as well, if you prefer. Just not (b) please :)

What do you think?

Reply:
After careful consideration, I think c) is best. If we get reports from confused users, we can always implement a)

Reply:
Awesome! :tada: